tests/unit/test_is_stuck.py

```diff
@@ -202,81 +239,75 @@ def test_is_stuck_repeating_action_error(
 
-        ipython_action_4 = IPythonRunCellAction(code='print("hello')
-        event_stream.add_event(ipython_action_4, EventSource.AGENT)
-        ipython_observation_4 = IPythonRunCellObservation(
-            content='print("hello\n       ^\nSyntaxError: unterminated string literal (detected at line 2)',
-            code='print("hello',
+        with patch('logging.Logger.warning'):
+            assert stuck_detector.is_stuck() is True
+
+    def test_is_not_stuck_incomplete_input_error(
+        self, stuck_detector: StuckDetector, event_stream: EventStream
+    ):
+        self._impl_syntax_error_events(
+            event_stream,
+            error_message='SyntaxError: incomplete input',
+            random_line=True,
         )
-        ipython_observation_4._cause = ipython_action_4._id
-        event_stream.add_event(ipython_observation_4, EventSource.USER)
 
-        # stuck_detector.state.history.set_event_stream(event_stream)
+        with patch('logging.Logger.warning'):
+            assert stuck_detector.is_stuck() is False
 
-        last_observations = [
-            ipython_observation_1,
-            ipython_observation_2,
-            ipython_observation_3,
-            ipython_observation_4,
-        ]
-        for observation in last_observations:
-            has_string = (
-                observation.content[-100:].find(
-                    'SyntaxError: unterminated string literal (detected at line'
-                )
-                != -1
-            )
-            assert has_string
-
-            string_is_last = (
-                len(
-                    observation.content.split(
-                        'SyntaxError: unterminated string literal (detected at line'
-                    )[-1]
-                )
-                < 10
-            )
-            assert string_is_last
+    def test_is_not_stuck_ipython_unterminated_string_error(
+        self, stuck_detector: StuckDetector, event_stream: EventStream
+    ):
+        self._impl_unterminated_string_error_events(event_stream, random_line=True)
 
-        with patch('logging.Logger.warning') as mock_warning:
+        with patch('logging.Logger.warning'):
+            assert stuck_detector.is_stuck() is False
+
+    def test_is_stuck_ipython_unterminated_string_error(
+        self, stuck_detector: StuckDetector, event_stream: EventStream
+    ):
+        self._impl_unterminated_string_error_events(event_stream, random_line=False)
+
+        with patch('logging.Logger.warning'):
             assert stuck_detector.is_stuck() is True
-            mock_warning.assert_called_once_with(
-                'Action, IPythonRunCellObservation loop detected'
-            )
 
     def test_is_stuck_ipython_syntax_error_not_at_end(
         self, stuck_detector: StuckDetector, event_stream: EventStream
     ):
+        # this test is to make sure we don't get false positives
+        # since the "at line x" is changing in between!
```

Comment:
🤔
FWIW, I wonder if requiring the error message to be identical isn't too restrictive now, hah (I'll see myself out lol). Please correct me if I'm not seeing right (looking at the diff), around line 190 in stuck.py it looks like the error too has to be entirely the same. I recall the stuck llm tries to make minor variations when it sees a syntax error, like it might try to run stuff starting with quotes or backquotes. It's still a syntax error, but it can be on another word or another line.

(FWIW I tend to think that this test was succeeding because the "syntax error" string had to be near the end of the line or something like that, not requiring an identical line number. I could be wrong though.)

Maybe that's all okay, since we're reducing from 4 to 3. 🤔 Three but identical might work well enough. The llm did seem to have a favorite thing to try lol, while variations of the code to run (e.g. when it tried running with quotes) was more rare. If this works in the scenario you found, we can go with it.

Reply:
The error doesn't have to be exactly the same, but the first line in those errors is like this:
`  Cell In[1], line 42`
So similar to the other syntax error, where a change of the line number "breaks" the loop, I added the first line to be the same in 3 observations. The first line does not contain the actual error, but the location.
Then it is up to the bottom 3 lines to match, basically (error type and 2 lines of our prompt).

I'll try to re-run a couple of instances from the Aider bench, from where I had that error collection from originally. :)